GraalVM-Truffle.md

```diff
@@ -1,3 +1,116 @@
+- It is built using the Truffle APIs.
+
+### Guest Language Interpreter
+
+- The Guest Language Interpreter is the runtime component that executes the guest language code.
+- It is compiled with a Java compiler from the Guest Language Implementation and the Truffle Framework.
+- Given a Guest Language Application, the Guest Language Interpreter executes the code and provides the necessary runtime support for the language features.
+  
+### Guest Language Application
+
+- The Guest Language Application is the code written in the guest language that is executed by the Guest Language Interpreter.
+- It can be any program or script that uses the features of the guest language.
+
+### HotSpotVM and GraalVM
+
+- The HotSpotVM is the Java Virtual Machine that executes the Guest Language Application.
+- It uses the GraalVM as a JIT compiler to optimize the execution of the guest language code.
+- The GraalVM applies advanced and speculative optimizations to improve performance [1,5].
+
+## GraalVM Compilation Process
+
+Given a JVM bytecode, you can use GraalVM to run and optimize it. The FMC diagram visualizes the JVM mode of GraalVM.
+
+![GraalVM Architecture](img/graal-fmc-Graal.drawio.png)
+
+### HotSpotVM
+
+- The HotSpotVM works as Java Virtual Machine (JVM) and executes Java applications using tiered compilation.
+- Initially, the interpreter runs the bytecode. As the code becomes "hot," it is compiled first with the client compiler (C1) and later with the server compiler (C2) [7].
+
+### Java Virtual Machine Compiler Interface (JVMCI)
+
+- The JVMCI is an API that allows any compiler to be used as a JIT compiler for C1 and C2.
+- This enables the integration of Graal's advanced optimization techniques into the JVM [5,7,8].
+
+### GraalVM
+
+- GraalVM is a high-performance runtime that includes a JIT compiler for optimizing Java bytecode [1,5].
+- It applies advanced and speculative optimizations.
+
+### GraphBuilder
+
+- The GraphBuilder is a component of the GraalVM that constructs from the JVM bytecode an intermediate representation (IR).
+- Other components of the GraalVM use this IR for further optimization and code generation [5].
+
+### GraalVM Backend
+
+- The GraalVM backend is responsible for generating machine code from the IR produced by the GraphBuilder.
+- It applies various optimizations to produce efficient native code.
+- It uses the Low Intermediate Representation (Low IR) to represent the code in a form that is closer to machine code [1,5].
+
+### Ideal Graph Visualizer (IGV)
+
+- The IGV is a tool that allows developers to visualize the IR of the code during different phases of the compilation process.
+- It provides insights into how the GraalVM optimizes the code and helps in debugging and performance tuning [3].
+
+Besides the JVM mode, GraalVM also supports the Native Image mode, which allows you to compile Java applications into native executables [4].
```

Comment:
Reference your already mentioned AOT paragraph?

Reply:
Uh yes! Great idea. Thank you 🥳